File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter06/ColorScrollWithViewModel/ColorScrollWithViewModel/RgbViewModel.cpp
#include "pch.h"
#include "RgbViewModel.h"

using namespace ColorScrollWithViewModel;

using namespace Platform;
using namespace Windows::UI;
using namespace Windows::UI::Xaml::Data;
using namespace Windows::UI::Xaml::Media;

RgbViewModel::RgbViewModel()
{
    this->color = ColorHelper::FromArgb(255, 0, 0, 0);
}

void RgbViewModel::Red::set(double value)
{
    if (red != value)
    {
        red = value;
        OnPropertyChanged("Red");
        Calculate();
    }
}

double RgbViewModel::Red::get()
{
    return red;
}

void RgbViewModel::Green::set(double value)
{
    if (green != value)
    {
        green = value;
        OnPropertyChanged("Green");
        Calculate();
    }
}

double RgbViewModel::Green::get()
{
    return green;
}

void RgbViewModel::Blue::set(double value)
{
    if (blue != value)
    {
        blue = value;
        OnPropertyChanged("Blue");
        Calculate();
    }
}

double RgbViewModel::Blue::get()
{
    return blue;
}

Color RgbViewModel::Color::get()
{
    return color;
}

void RgbViewModel::Calculate()
{
    this->color = ColorHelper::FromArgb(255, unsigned char(this->Red), 
                                             unsigned char(this->Green), 
                                             unsigned char(this->Blue));
    OnPropertyChanged("Color");
}

void RgbViewModel::OnPropertyChanged(String^ propertyName)
{
    PropertyChanged(this, ref new PropertyChangedEventArgs(propertyName));
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter06/ColorScrollWithViewModel/ColorScrollWithViewModel/RgbViewModel.cs
using System.ComponentModel;        // for INotifyPropertyChanged
using Windows.UI;                   // for Color

namespace ColorScrollWithViewModel
{
    public class RgbViewModel : INotifyPropertyChanged
    {
        double red, green, blue;
        Color color = Color.FromArgb(255, 0, 0, 0);

        public event PropertyChangedEventHandler PropertyChanged;

        public double Red
        {
            set
            {
                if (red != value)
                {
                    red = value;
                    OnPropertyChanged("Red");
                    Calculate();
                }
            }
            get
            {
                return red;
            }
        }

        public double Green
        {
            set
            {
                if (green != value)
                {
                    green = value;
                    OnPropertyChanged("Green");
                    Calculate();
                }
            }
            get
            {
                return green;
            }
        }

        public double Blue
        {
            set
            {
                if (blue != value)
                {
                    blue = value;
                    OnPropertyChanged("Blue");
                    Calculate();
                }
            }
            get
            {
                return blue;
            }
        }

        public Color Color
        {
            protected set
            {
                if (color != value)
                {
                    color = value;
                    OnPropertyChanged("Color");
                }
            }
            get
            {
                return color;
            }
        }

        void Calculate()
        {
            this.Color = Color.FromArgb(255, (byte)this.Red, (byte)this.Green, (byte)this.Blue);
        }

        protected void OnPropertyChanged(string propertyName)
        {
            if (PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
        }
    }
}
